### PR TITLE
Dev timed animation fix

### PR DESCRIPTION
**Pull request by danhalldubit**
If a TimedAnimation was given a duration of <= 0 in the constructor, the Duration property setter called FastForward and invoked the OnComplete callback without having Play called.

### CODE REVIEW

File: Tween/TimedAnimation.cs [hunk 1]
```diff
@@ -68,8 +56,8 @@ public float Duration
 
 		public override bool IsValid { get { return TargetObject != null || TargetRectTransform != null; } }
 
-		protected bool IsComplete { get { return IsReversed ? Progress <= 0 : Progress >= 1.0f; } }
-		protected float Progress { get { return Mathf.Clamp01(CurrentTime / Duration); } }
+		protected bool IsComplete { get { return IsReversed ? Progress <= 0f : Progress >= 1f; } }
+		private float Progress { get { return Mathf.Clamp01(CurrentTime / Duration); } }
```
Comment (AndrewThompson85) [inline Tween/TimedAnimation.cs]: Is there any reason not to make this public ?

Comment (Rollo-Liu) [reply Tween/TimedAnimation.cs]: Yes.
1. It was protected and nobody was using it;
2. `Duration` can be 0 and you will have a "divide by 0" issue -- making it private will decrease the chance spreading this issue.

File: Tween/TimedAnimation.cs [hunk 2]
```diff
@@ -68,8 +56,8 @@ public float Duration
 
 		public override bool IsValid { get { return TargetObject != null || TargetRectTransform != null; } }
 
-		protected bool IsComplete { get { return IsReversed ? Progress <= 0 : Progress >= 1.0f; } }
-		protected float Progress { get { return Mathf.Clamp01(CurrentTime / Duration); } }
+		protected bool IsComplete { get { return IsReversed ? Progress <= 0f : Progress >= 1f; } }
```
[the remark this answers is not in the document]
Comment (danhalldubit) [reply Tween/TimedAnimation.cs]: Yeah it's only used in the Update of this class, I'll make it private

File: Tween/TimedAnimation.cs [hunk 3]
```diff
@@ -101,14 +89,24 @@ public override void Play(Action onComplete)
...
+				// A little guard for unnecessary adding to the update list
+				if (!IsPlaying)
+				{
+					AnimationDriver.Add(Update);
+				}
+
+				base.Play(onComplete);
+				Refresh(CurrentTime = IsReversed ? Duration : 0f);
+			}
+			else
```
[the remark this answers is not in the document]
Comment (danhalldubit) [reply Tween/TimedAnimation.cs]: Done

**Comment by danhalldubit** [comment]
Just added an optional `OnAbort` callback for animations. Previously when an animation aborted itself because it became invalid, it would invoke the `OnComplete` anyway, mainly to allow `SequencedAnimation`s to continue to the next animation.

This wasn't really expected behaviour from a user standpoint, and changing it has the added bonus of not needing to manually cache the animation, and abort it on `Destroy` if the `OnComplete` callback could cause an exception.

**Pull request merged**